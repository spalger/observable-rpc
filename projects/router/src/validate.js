import { RpcError } from '@observable-rpc/core'

export function validate(toValidate, schema, options = {}) {
  const { context, desc = 'Validation' } = options

  const { value, error } = schema.validate(toValidate, {
    abortEarly: false,
    stripUnknown: true,
    context,
  })

  if (error) {
    throw RpcError.from(error, {
      statusCode: 401,
      message: `${desc} failure`,
    })
  }

  return value
}
