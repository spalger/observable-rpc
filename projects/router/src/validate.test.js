import { Joi } from './joi'
import { validate } from './validate'

it('returns the validated data, including defaults from schema', () => {
  expect(validate({}, Joi.object().keys({ foo: Joi.default('bar') }))).toEqual({
    foo: 'bar',
  })
})

it('throws if input does not match schema', () => {
  expect(() => {
    validate({}, Joi.object({ foo: Joi.valid('bar').required() }))
  }).toThrowErrorMatchingSnapshot()
})

it('includes validation description if error message if supplied', () => {
  expect(() => {
    validate(
      {},
      Joi.object({ foo: Joi.valid('bar').required() }),
      'Foo validation'
    )
  }).toThrowErrorMatchingSnapshot()
})

it('throws a RpcError 401 error', () => {
  let thrown
  try {
    validate({}, Joi.object({ foo: Joi.valid('bar').required() }))
  } catch (error) {
    thrown = error
  }

  expect(thrown).toHaveProperty('isRpcError', true)
  expect(thrown).toHaveProperty('statusCode', 401)
})
