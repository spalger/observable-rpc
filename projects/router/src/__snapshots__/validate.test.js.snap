// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`includes description in error message if supplied 1`] = `"Foo validation failure: child \\"foo\\" fails because [\\"foo\\" is required]"`;

exports[`throws if input does not match schema 1`] = `"Validation failure: child \\"foo\\" fails because [\\"foo\\" is required]"`;
